src/hooks/useQuestionnaireResponseDraft.ts

```diff
@@ -0,0 +1,268 @@
+}
+
+export const useQuestionnaireResponseDraft = (
+    props: QuestionnaireResponseDraftProps,
+): QuestionnaireResponseDraftResponse => {
+    const { autoSave = false, qrDraftServiceType = 'local', subject, questionnaireId, questionnaireResponse } = props;
+
+    const [draftInfoMessage, setDraftInfoMessage] = useState<string | undefined>();
+    const draftKeyRef = useRef<string | undefined>();
+
+    const [response, manager] = useService<WithId<QuestionnaireResponse> | undefined>(async () => {
+        const questionnaireRD = await getFHIRResources<Questionnaire>('Questionnaire', {
+            id: questionnaireId,
+            _elements: ['id', 'meta'].join(','),
+        });
+
+        if (isFailure(questionnaireRD)) {
+            return failure(t`Questionnaire not found`);
+        }
+
+        const questionnaire = extractBundleResources(questionnaireRD.data).Questionnaire[0];
+
+        draftKeyRef.current =
+            qrDraftServiceType === 'server'
+                ? makeServerDraftKey(questionnaireResponse?.id)
+                : makeLocalStorageDraftVersionedKey({
+                      subject,
+                      questionnaire,
+                      questionnaireResponse,
+                  });
+        const draftQRRD = await loadQuestionnaireResponseDraft(draftKeyRef.current, qrDraftServiceType);
+
+        if (isFailure(draftQRRD)) {
+            return success(questionnaireResponse);
+        }
+
+        const resultQR = {
+            ...questionnaireResponse,
+            ...draftQRRD.data,
+        };
+
+        setDraftInfoMessage(
+            t`Draft was successfully loaded from ${qrDraftServiceType === 'local' ? 'local storage' : 'FHIR server'}`,
+        );
+
+        return success(resultQR);
+    }, [questionnaireId]);
+
+    const saveDraft = useCallback(
+        async (questionnaireResponse: QuestionnaireResponse): Promise<RemoteDataResult<QuestionnaireResponse>> => {
+            if (!autoSave || !questionnaireResponse) {
+                const reason = !autoSave
+                    ? t`Auto save is disabled`
+                    : !questionnaireResponse
+                    ? t`Questionnaire response is required`
+                    : t`Form data is required`;
+                return failure(reason);
+            }
+
+            const draftQRRD = await saveQuestionnaireResponseDraft({
+                keyId: draftKeyRef.current,
+                questionnaireResponse,
+                qrDraftServiceType,
+            });
+
+            return draftQRRD;
+        },
+        [autoSave, qrDraftServiceType],
+    );
+
+    const isRunningDebouncedSaveDraftRef = useRef(false);
+    const debouncedSaveDraftRef = useRef<ReturnType<typeof _.debounce> | null>(null);
+
+    useEffect(() => {
+        debouncedSaveDraftRef.current = _.debounce(async (questionnaireResponse: QuestionnaireResponse) => {
+            if (isRunningDebouncedSaveDraftRef.current) {
+                return;
+            }
+
```

Comment:
It must be checked here, not in saveDraft. saveDraft is generic function, it should not know about autosave
```
if (!autoSave ) {
    return;
}
```